delphi/translators/for2py/rectify.py

```diff
@@ -463,31 +826,117 @@ def handle_tag_statement(self, root, parElem):
+
+                        if "parent-scope" not in current.attrib:
+                            current.attrib['skip-collect'] = "true"
+                        self.goto_target_lbl_after.append(target_lbl)
+                        self.collect_stmts_after_goto = True
+                        self.label_after = True
+                else:
+                    if target_lbl in self.label_lbl_for_before:
+                        assert traverse > 1, "Reconstruction cannot happen in the first traverse"
+                        if self.label_before:
+                            self.reconstruct_before_case_now = True
+                        return
             else:
-                print(
-                    f'In self.handle_tag_statement: "{child.tag}" not handled'
-                )
+                assert False, f'In handle_tag_statement: "{child.tag}" not handled'
+
+        # Statement collector (1st traverse)
+        if traverse == 1:
+            if self.label_before and not self.label_after:
+                # Since we do not want to extract the stop statement from
+                # that is not a main body, check it before extraction
+                if "has-stop" not in current.attrib:
+                    current.attrib['goto-move'] = "true"
+                    self.statements_to_reconstruct_before['stmts-follow-label'].append(current)
+                else: 
+                    if (
+                        parent.tag == "body" 
+                        and parent.attrib['parent'] == "program"
+                    ):
+                        self.statements_to_reconstruct_before['stmts-follow-label'].append(current)
+            elif self.label_after:
+                if self.collect_stmts_after_goto:
+                    current.attrib['goto-remove'] = "true"
+                    if "has-stop" not in current.attrib:
+                        self.statements_to_reconstruct_after['stmts-follow-goto'].append(current)
+                    else:
+                        if (
+                            parent.tag == "body"
+                            and parent.attrib['parent'] == "program"
+                        ):
+                            self.statements_to_reconstruct_after['stmts-follow-goto'].append(current)
+
+                    if "goto-stmt" in current.attrib:
+                        self.stmts_after_goto['goto-stmts'].append(current.attrib['lbl'])
+                    elif "target-label-statement" in current.attrib:
+                        self.stmts_after_goto['labels'].append(current.attrib['label'])
+                        
+                elif self.collect_stmts_after_label:
+                    current.attrib['goto-remove'] = "true"
+                    self.statements_to_reconstruct_after['stmts-follow-label'].append(current)
+                    
+                    if (
+                        parent.tag == "body"
+                        and parent.attrib['parent'] == "program"
```

Comment:
Please parenthesize this expression so that the relationship between the AND and OR is obvious.